torchdata/datapipes/iter/util/prefetch.py

```diff
@@ -0,0 +1,214 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import threading
+
+import torch
+import torch.distributed as dist
+
+from collections import deque
+from concurrent.futures import ThreadPoolExecutor, TimeoutError, Future
+from dataclasses import dataclass
+from functools import partial
+from typing import Callable, Deque, Iterator, Optional, TypeVar
+
+from torchdata._constants import default_timeout_in_s
+from torchdata.datapipes import functional_datapipe
+from torchdata.datapipes.iter import IterDataPipe
+
+T_co = TypeVar("T_co", covariant=True)
+
+
+__all__ = ["Expected", "FullSyncIterDataPipe", "PrefetchTimeoutError"]
+
+
+class PrefetchTimeoutError(RuntimeError):
+    def __init__(self, timeout: int) -> None:
+        super().__init__(f"Fail to fetch data within {timeout} seconds")
+
+
+class _EndOfPrefetch:
+    ...
+
+
+@dataclass
+class Expected:
+    r"""
+    Expected data provided to callback function in ``_PrefetchExecutor``.
+    """
+    index: int
+    error: Optional[BaseException] = None
+
+    def has_error(self) -> bool:
+        return self.error is not None
+
+
+class _PrefetchExecutor:
```

Comment:
This executor takes reference from the internal implementation: https://fburl.com/code/7dk6mvs4
On the top of the implementation, I added `prefetch_size` to make sure it can work with `Prefetch` in the future.